### PR TITLE
[Flow] Improve slice raising to handle dynamic and unit dims

### PR DESCRIPTION
**Pull request by qedawkins**
This enables the following raising cases:
  1. Dynamic dimensions able to be statically analyzed as matching the associated dimension in the output.
  2. Unit dimensions indexed by 0.

Dynamic dimensions are compared using the ValueBounds interface, however this currently has no pattern for analyzing tensor.expand/collapse shape ops, making this pattern interact poorly with FoldUnitExtentDims.

Also cleans up some of the test cases for slice raising.

### CODE REVIEW

File: compiler/src/iree/compiler/Dialect/Flow/Transforms/RaiseSpecialOps.cpp [hunk 1]
```diff
@@ -245,42 +262,50 @@ tryRaiseToExtractSlice(AffineMap inputIndexingMap, AffineMap outputIndexingMap,
...
-                                                  currOutDim);
-        sizes.push_back(dim.getResult());
-      } else {
-        sizes.push_back(rewriter.getI64IntegerAttr(outShape[currOutDim]));
-      }
-      ++currOutDim;
-      continue;
-    }
-    // Assume that the constant access is a rank reducing access.
+    // Constant accesses can either be rank reducing or an access into a unit
```
Comment (MaheshRavishankar) [inline compiler/src/iree/compiler/Dialect/Flow/Transforms/RaiseSpecialOps.cpp:288]: This is actually limiting the cases covered? The constant access does not have to be for dimensions that are unit dim. You could have a constant access where the source of the unit dim could be anything. Then the constant (or any ssa value for that matter that is not related the Linalg op loops become the offset, and the size is 1.

I'd like to take a bit of a further look into this.

**Review by MaheshRavishankar — CHANGES_REQUESTED** [review]
I need to take a look a bit more here.

File: compiler/src/iree/compiler/Dialect/Flow/Transforms/RaiseSpecialOps.cpp (hunk 1 above)
Comment (qedawkins) [reply compiler/src/iree/compiler/Dialect/Flow/Transforms/RaiseSpecialOps.cpp:288]: I think the comment here is a bit misleading. It does have to either be a unit dim or rank reducing on the output though due to the constraints of a linalg op. If not one of those then it is a broadcast and thus can't be raised, but I think that should be captured properly. That said, I might not have thought through it properly and will come back to it later. I'm also starting to think that this intermediate linalg approach is nice for the cases that work, but there are certain classes of slices (exactly when we have an arbitrary computation of an index/size for slicing) that warrant a direct lowering. I'd like to see real world use cases for those before adding raisings for them though.

I'll add a test case for broadcast to make sure it isn't erroneously raised though.
